[Android] Fix misused 'continue' statements.

adamradocz opened the pull request:
**Description of the Change**
The `continue` statement has been used wrongly a couple of times. I changed to `break` or deleted them, but it changes the logic of the loops, so please check thoroughly.

**Alternate Designs**
Some `for + break` loop could have changed to `while`.

**Release Notes**
[Android] Minor improvement.

adamradocz commented:
`/android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.java`

```
// figure out what index PID has
     	String [] headers = processLinesAr[0].split("[\\s]+");
     	Integer PidIndex = 1;
    	for (int x = 0; x < headers.length; x++) {
    		if(headers[x].equals("PID")) {
    			PidIndex = x;
    			break;
    		}
    	}
```
I don't understand why Integer PidIndex gets an initial 1 value.
Isn't that should be less than 0 or null, because we pass x to PidIndex and x can be 0 and 1 too. Or the for should start from 2, because the first two loop is unnecessary.

Hope I was clear :)

adamradocz commented:
`/android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.java`

Line 832:
```
for(int y = 1; y < processLinesAr.length; y++) {
    		Boolean found = false;
    	    Boolean found = false;
    	    String [] comps = processLinesAr[y].split("[\\s]+");
    	    for(String arg: comps) {
    	    	if(arg.equals(processName)) {
     	    		if(Logging.DEBUG) Log.d(Logging.TAG,"getPidForProcessName(): " + processName + " found in line: " + y);
     	    		found = true;
     	    	}
     	    }
     	    if(found) {
 	    	    try{
	    	    	pid = Integer.parseInt(comps[PidIndex]);
	        	    if(Logging.ERROR) Log.d(Logging.TAG,"getPidForProcessName(): pid: " + pid); 
	    	    }catch (NumberFormatException e) {if(Logging.ERROR) Log.e(Logging.TAG,"getPidForProcessName(): NumberFormatException for " + comps[PidIndex] + " at index: " + PidIndex);}
	    	    break;
    	    }
    	}
```

I replaced the `continue` statement to `break` after the `catch` tag, but I think it was intended in the `foreach` loop under the `found = true;`. Am I right?

brevilo commented on android/BOINC/app/src/main/java/edu/berkeley/boinc/ProjectsFragment.java:
```diff
@@ -158,7 +158,7 @@ private void updateData(ArrayList<Project> latestRpcProjectsList, AcctMgrInfo ac
 			if(data.get(x).isMgr) {
 				mgrIndex = x;
-				continue;
+				break;
```
The question here is: can there be more than one AM entry? If so, which one should be recorded, the last (original) or the first (your version)? Given the context I tend to accept your change.

AenBleidd replied:
I think changing `continue` to `break` will break the logic here. I'd rather just remove `continue` here.
BTW, I don't fully understand the purpose of this function at all.
In case of mane AMs present in the list current implementation will overwrite the last AM entry.
New proposed implementation will overwrite first only.
I think more detailed investigation is necessary here to understand the purpose and desired behavior of this function

brevilo replied:
Ok, so the decision is to not optimize this right now, but to flag this as an open question (providing how we think it should be written). FTR: removing the `continue` doesn't change the semantics.

brevilo commented on android/BOINC/app/src/main/java/edu/berkeley/boinc/PrefsFragment.java:
```diff
@@ -219,7 +219,6 @@ private void updateBoolPref(int ID, Boolean newValue) {
 		for (PrefsListItemWrapper item: data) {
 			if(item.ID == ID){
 				((PrefsListItemWrapperBool) item).setStatus(newValue);
-				continue;
```
Ok.

brevilo commented on android/BOINC/app/src/main/java/edu/berkeley/boinc/PrefsFragment.java:
```diff
@@ -231,7 +230,6 @@ private void updateValuePref(int ID, Double newValue) {
 		for (PrefsListItemWrapper item: data) {
 			if(item.ID == ID){
 				((PrefsListItemWrapperValue) item).status = newValue;
-				continue;
```
Ok.

brevilo commented on android/BOINC/app/src/main/java/edu/berkeley/boinc/ProjectsFragment.java:
```diff
@@ -185,7 +185,7 @@ private void updateData(ArrayList<Project> latestRpcProjectsList, AcctMgrInfo ac
+					break;
 				}
 			}
 			if(index == null) { // result is new, add
```
I think all the comment/logging references to "result" or wrong here and likely a copy 'n' paste issue. Shouldn't all of those occurrences refer to "project" instead??? This doesn't really help understanding the original code's purpose.

adamradocz replied:
It looks fine to me.

adamradocz replied:
Seems fine to me.

brevilo replied:
How? This code section doesn't deal with results at all but with projects. Just to be clear, this not about your change, it's an additional change that would be good to get in, since you're at it anyway.

AenBleidd replied:
I agree with @brevilo. I believe comments and log messages should be changed from `result` to `project`

adamradocz replied:
Fixed.

brevilo commented on android/BOINC/app/src/main/java/edu/berkeley/boinc/ProjectsFragment.java:
```diff
@@ -185,7 +185,7 @@ private void updateData(ArrayList<Project> latestRpcProjectsList, AcctMgrInfo ac
 				if(rpcResult.master_url.equals(data.get(x).id)) {
 					index = x;
-					continue;
+					break;
```
Should be ok as well as finding a single match seems to be the intention here.

AenBleidd replied:
Same as above. Need more info

adamradocz replied:
I disagree. The comment above says, "Check whether this Result is new". If we pass the `index` a value, the `if (index == null)` is already false, so we don't need to find more `X`s.

AenBleidd replied:
@adamradocz, I agree with you but the open question for me now it why this function is implemented this way. Maybe the current behavior is incorrect at whole but changing the logic without full understanding of the situation in worse than just leaving it as is for a while.
Please understand me correctly, I also want to make this product better, as all of us I believe.
And thank you for your hard job!

brevilo replied:
I think the whole codebase shows that the original author had no clear understanding of what `continue` means. So we have to take that into account while trying to determine the **originally intended** functionality.

adamradocz replied:
I made the changes that @AenBleidd proposed.

brevilo replied:
Same as above: removing `continue` doesn't change the semantics and the proposed solution is visible, yet disabled until further clarification.

brevilo commented on android/BOINC/app/src/main/java/edu/berkeley/boinc/ProjectsFragment.java:
```diff
@@ -207,7 +207,7 @@ private void updateData(ArrayList<Project> latestRpcProjectsList, AcctMgrInfo ac
 				if(listItem.id.equals(rpcResult.master_url)) {
 					found = true;
-					continue;
+					break;
```
ok

brevilo commented on android/BOINC/app/src/main/java/edu/berkeley/boinc/ProjectsFragment.java:
```diff
@@ -225,7 +225,6 @@ private void updateData(ArrayList<Project> latestRpcProjectsList, AcctMgrInfo ac
 						project.addServerNotice(serverNotice);
 						noticeFound = true;
 						mappedServerNotices++;
-						continue;
```
Ok.

brevilo commented on android/BOINC/app/src/main/java/edu/berkeley/boinc/TasksFragment.java:
```diff
@@ -122,7 +122,7 @@ private void updateData(ArrayList<Result> newData) {
 				if(rpcResult.name.equals(data.get(x).id)) {
 					index = x;
-					continue;
+					break;
```
Ok.

brevilo commented on android/BOINC/app/src/main/java/edu/berkeley/boinc/TasksFragment.java:
```diff
@@ -142,7 +142,7 @@ private void updateData(ArrayList<Result> newData) {
 				if(listItem.id.equals(rpcResult.name)) {
 					found = true;
-					continue;
+					break;
```
Ok.

brevilo commented on android/BOINC/app/src/main/java/edu/berkeley/boinc/client/ClientInterfaceImplementation.java:
```diff
@@ -184,18 +184,16 @@ public Boolean attachProject(String url, String projectName, String authenticato
 				if(project.master_url.equals(url)) {
-					match = true;
-					continue;
+					return true;
```
Ok (and all related changes)

brevilo commented on android/BOINC/app/src/main/java/edu/berkeley/boinc/client/ClientStatus.java:
```diff
@@ -589,7 +589,7 @@ private void parseComputingStatus() {
 						if(task.active_task) { // this result has corresponding "active task" in RPC XML
 							activeTask = true;
-							continue; // amount of active tasks does not matter.
+							break; // amount of active tasks does not matter.
```
Certainly OK (as the comment says!)

brevilo commented on android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.java:
```diff
@@ -823,27 +823,27 @@ private Integer getPidForProcessName(String processName) {
     	    	if(arg.equals(processName)) {
     	    		if(Logging.DEBUG) Log.d(Logging.TAG,"getPidForProcessName(): " + processName + " found in line: " + y);
     	    		found = true;
+                    break;
```
Please use the same indentation level or adjust the previous two lines as well. Otherwise ok.

brevilo commented on android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.java:
```diff
@@ -823,27 +823,27 @@ private Integer getPidForProcessName(String processName) {
 	    	    	pid = Integer.parseInt(comps[PidIndex]);
 	        	    if(Logging.ERROR) Log.d(Logging.TAG,"getPidForProcessName(): pid: " + pid); 
 	    	    }catch (NumberFormatException e) {if(Logging.ERROR) Log.e(Logging.TAG,"getPidForProcessName(): NumberFormatException for " + comps[PidIndex] + " at index: " + PidIndex);}
-	    	    continue;
```
Ok.

brevilo commented on android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.java:
```diff
@@ -823,27 +823,27 @@ private Integer getPidForProcessName(String processName) {
     		if(headers[x].equals("PID")) {
     			PidIndex = x;
-    			continue;
+    			break;
```
Ok.

brevilo commented:
> I don't understand why `Integer PidIndex` gets an initial `1` value.

That doesn't make sense indeed. So, yes, I think it should be initialized with `-1`.
And it's worse than just that: the following debug log statement as well as the next code block don't check if the `PID` column was found at all. Lines 829 to 848 should be wrapped in a conditional that executes `if(PidIndex != -1)`.

brevilo commented:
> I replaced the `continue` statement to `break` after the `catch` tag, but I think it was intended in the `foreach` loop under the `found = true;`. Am I right?
> 
> **Update:** I moved the `break` from the `if` to the `foreach` statement. If OK, I'll fix the indentation.

Sorry, I'm not what exactly is being referred to here: the snippet your posted or the change you committed? Please clarify.

adamradocz commented:
> > I don't understand why `Integer PidIndex` gets an initial `1` value.
> 
> That doesn't make sense indeed. So, yes, I think it should be initialized with `-1`.
> And it's worse than just that: the following debug log statement as well as the next code block don't check if the `PID` column was found at all. Lines 829 to 848 should be wrapped in a conditional that executes `if(PidIndex != -1)`.

Lines between 829 and 850 should be wrapped in a conditional statement.

adamradocz commented:
> > I replaced the `continue` statement to `break` after the `catch` tag, but I think it was intended in the `foreach` loop under the `found = true;`. Am I right?
> > **Update:** I moved the `break` from the `if` to the `foreach` statement. If OK, I'll fix the indentation.
> 
> Sorry, I'm not sure what exactly is being referred to here: the snippet you posted or the change you committed? Please clarify.
> 
> I think the `break` you [added](https://github.com/BOINC/boinc/pull/2898/files/facb70ee9887e88ebfa003563823a498d04a6d1c#diff-1b828896f8d964b096d9187399c8e0e8R839) to the for-each is as justified as all the others you added elsewhere (here: stop when matching process found). But I also think that it shouldn't just replace the `continue` you [removed](https://github.com/BOINC/boinc/pull/2898/files/facb70ee9887e88ebfa003563823a498d04a6d1c#diff-1b828896f8d964b096d9187399c8e0e8L846). That instead should also be replaced by a `break` to stop the search when the process' PID was found.
I added back the 2nd `break`, but honestly I don't like the two `break`s here. It makes the `for` loop unreadable, but I will leave it for now.

brevilo commented on android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.java:
```diff
@@ -819,35 +819,48 @@ private Integer getPidForProcessName(String processName) {
+					if (Logging.DEBUG)
+						Log.d(Logging.TAG, "getPidForProcessName(): " + processName + " found in line: " + y);
+					found = true;
+					break;
```
Since you find it confusing, just add a comment like "break out of inner for-each (comps) loop"

adamradocz replied:
Fixed.

brevilo commented on android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.java:
```diff
@@ -819,35 +819,48 @@ private Integer getPidForProcessName(String processName) {
+					if (Logging.ERROR)
+						Log.e(Logging.TAG, "getPidForProcessName(): NumberFormatException for " + comps[PidIndex] + " at index: " + PidIndex);
+
+					break;
```
Since you find it confusing, just add a comment like "break out of outer for (processLinesAr.length) loop"

brevilo replied:
Also, I think this `break` should go after the `try/catch` block as the outer for loop should be exited after the process was found, regardless of whether the PID could be extracted or not.

adamradocz replied:
True. Fixed.

brevilo commented on android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.java:
```diff
@@ -819,35 +819,48 @@ private Integer getPidForProcessName(String processName) {
-		if(Logging.DEBUG) Log.d(Logging.TAG,"getPidForProcessName(): PID at index: " + PidIndex + " for output: " + processLinesAr[0]);
-    	
+
+		if (PidIndex == -1)
```
Please review the indentation of all lines between 829 and 864. At least GitHub shows things are really off right now. Thanks.

adamradocz commented on android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.java:
```diff
@@ -819,35 +819,48 @@ private Integer getPidForProcessName(String processName) {
-    	
+
+		if (PidIndex == -1)
+			return null;
```
Fixed.

adamradocz commented:
Done.

brevilo commented on android/BOINC/app/src/main/java/edu/berkeley/boinc/client/Monitor.java:
```diff
@@ -827,38 +827,35 @@ private Integer getPidForProcessName(String processName) {
+    			} catch (NumberFormatException e) {
+    				if (Logging.ERROR) Log.e(Logging.TAG, "getPidForProcessName(): NumberFormatException for " + comps[PidIndex] + " at index: " + PidIndex);
+    			}
+    			break;// Break out of inner for (processLinesAr) loop
```
This breaks out of the *outer* loop

brevilo commented:
Waiting for CI to finish. Then I'll merge...
